### PR TITLE
Add some documentation on the wire format.

### PR DESCRIPTION
**Pull request by mikehearn**

### CODE REVIEW

File: docs/source/wire-format.rst [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+===========
+
+This document describes the Corda wire format. With the following information and an implementation of the AMQP/1.0
+specification, you can read Corda serialised binary messages. An example implementation of AMQP/1.0 would be Apache
+Qpid Proton, or Microsoft AMQP.NET Lite.
+
+Header
+------
+
+All messages start with the byte sequence ``corda\1\0``, that is, the string "corda" followed by a one byte and then a
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: The header is actually 8 bytes.  So there's one more byte.  It's defined in `SectionId`.  In reality, this should always be zero (currently) if you are speaking to the node (either via RPC or P2P).  Again, abort if it is not zero.

File: docs/source/wire-format.rst [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+AMQP/1.0 (which is quite different to AMQP/0.9) is a standardised binary encoding scheme comparable to but more advanced
+than Google protocol buffers. It specifies encodings for several 'primitive' types: numbers, strings, UUIDs, timestamps
+and symbols (these can be thought of as enum entries). It also defines how to encode maps, lists and arrays. The difference
+between the latter two is that arrays always contain a single type, whereas lists can contain elements of different types.
+An AMQP byte stream is simply a repeated series of elements.
+
+So far, so standard. However AMQP goes further than most such tagged binary encodings by including the concept of
+'described types'. This is a way to impose an application-level type system on top of the basic "bags of elements"
+that low-level AMQP gives you. Any element in the stream can be prefixed with a *descriptor*, which is either a string
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: Is it worth providing a link to the Oasys docs:

Described types: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#doc-idp38080

Composite types (for use below): http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#doc-idp108672

Restricted types (for use below): http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#doc-idp125312

File: docs/source/wire-format.rst [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+1. The top level message and is described using a string (symbolic) descriptor.
+2. A *SCHEMA* record.
+3. A *TRANSFORM_SCHEMA* record.
+
+The transform schema will usually be empty - it's used to describe how a data structure has evolved over time, so
+making it easier to map to old/new code.
+
+The *SCHEMA* record always contains a single element, which is itself another list containing *COMPOSITE_TYPE* records.
+Each *COMPOSITE_TYPE* record describes a single app-level type and has the following members:
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: Maybe worth referencing the Oasys docs (links above) for composite and restricted types.  We basically mirror the XML documented there, but in AMQP binary form.

**Comment by mikehearn** [comment]
@rick-r3 Thanks for the review, good ideas.

I've pushed fixes and also added some more docs, plus an example of how it all maps to the wire.

File: docs/source/wire-format.rst [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+   having data on the ledger later on that's hard to read due to loss of (old versions of) applications. The intention is
+   that a mix of compression and separating the schema parts out when both sides already agree on what they are will return
+   most of the lost efficiency.
+
+Descriptors
+-----------
+
+Serialised messages use described types extensively. There are two types of descriptor:
+
+1. 64 bit code. In Corda, the top 16 bits are always equal to 0xc562 which is R3's IANA assigned enterprise number. The
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: top 32 bits

File: docs/source/wire-format.rst [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+2. Type: string
+3. Requires: list of string
+4. Default: nullable string
+5. Label: nullable string
+6. Mandatory: boolean
+7. Multiple: boolean
+
+The meaning of these are defined in the AMQP specification. The type string is a Java class name *with* generic parameters.
+
+The other parts of the schema map to the AMQP XML schema spec in the same straightforward manner.
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: spec -> specification

File: docs/source/wire-format.rst [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+Mapping JVM classes to composite types
+--------------------------------------
+
+Corda does not need or use a separate schema definition language. Instead, source code is used as a way to define schemas
+via regular class definitions in any statically typed JVM-bytecode targeting language. This specification will thus
+frequently to types whose only definitions are found in the Corda source code: these definitions are canonical and not
```
Comment (rick-r3) [inline docs/source/wire-format.rst]: Something is missing.  "This specification will thus frequently to types whose only definitions are found in the Corda source code"

"refer" missing ?

**Comment by shamsasari** [comment]
This PR seems abandoned @mikehearn.

**Comment by mikehearn** [comment]
@shamsasari Believe it or not, I'm getting back to it today!

**Comment by mikehearn** [comment]
@rick-r3 Fixed your review comments, PTAL!

**Pull request merged**